### DISCUSSION TITLE
web / api roles and permissions headache!

### DISCUSSION THREAD

**Discussion by tomekkicaj700** [spatie/laravel-permission#2421] [Q&A] [answered]
Hi Guys!

I cannot figure this out.

In my seeder I created:

```
    Permission::findOrCreate(Permissions::CAN_MANAGE_USERS, "api");
    Permission::findOrCreate(Permissions::CAN_MANAGE_USERS, "web");
    Permission::findOrCreate(Permissions::CAN_MANAGE_WELDERS, "api");
    Permission::findOrCreate(Permissions::CAN_MANAGE_WELDERS, "web");
```

I created ADMIN roles for `api `and `web`:

```
   Role::create([
      "guard_name" => "api",
      "name" => Roles::ADMIN,
    ])->givePermissionTo([
      Permissions::CAN_MANAGE_USERS,
      Permissions::CAN_MANAGE_WELDERS,
    ]);

    Role::create([
      "guard_name" => "web",
      "name" => Roles::ADMIN,
    ])->givePermissionTo([
      Permissions::CAN_MANAGE_USERS,
      Permissions::CAN_MANAGE_WELDERS,
    ]);
```
I assign user 1 the ADMIN role:

`User::find(1)->assignRole(Roles::ADMIN);`

So, my user 1 has ADMIN role which has a separate set of permissions for guards: `web `and `api `

Now , in my unit tests:

```
1.
    $this->actingAs(User::find(1), "api");
    dd(
      auth()
        ->user()
        ->can(Permissions::CAN_MANAGE_WELDERS)
    );
returns FALSE

2.
    $this->actingAs(User::find(1), "web");
    dd(
      auth()
        ->user()
        ->can(Permissions::CAN_MANAGE_WELDERS)
    );
returns TRUE
```

I don't understand why test 1 fails ???
I'm really helpless, Spatie manual does't help, it just mentions 2 sets of permissions for `api `and `web `but it's what I just did, but it doesn't work.

Thanks!

**Comment by parallels999** [comment]
> `User::find(1)->assignRole(Roles::ADMIN);`

That only assign `Roles::ADMIN` on `web`, not on `api`

**Reply by tomekkicaj700** [reply]
Then how do I make user 1 to have ADMIN role for both web and api guard?

**Reply by parallels999** [reply]
@drbyte there was a PR trying to make possible assignment on different guards https://github.com/spatie/laravel-permission/pull/2291, but it was a big breaking change
how about this idea:
```php
$user = User::find(1);
$user->usingGuard('web')->assignRole(Roles::ADMIN);
$user->usingGuard('api')->assignRole(Roles::ADMIN);
```
on `usingGuard`, it changes a property on `$user`, so 
https://github.com/spatie/laravel-permission/blob/e64d2f9fd510dc10174d66d170f1eb09efa45530/src/Guard.php#L16-L26
```php
public static function getNames($model): Collection 
 { 
     $class = is_object($model) ? get_class($model) : $model; 
  
     if (is_object($model)) { 
         if($model->usingGuard) {
             $guardName = $model->usingGuard;
         } else if (\method_exists($model, 'guardName')) { 
             $guardName = $model->guardName(); 
         } else { 
             $guardName = $model->getAttributeValue('guard_name'); 
         } 
     }
```
the idea can be improved, but I hope I have explained myself

**Reply by drbyte** [reply]
@parallels999 I see your point. 
But I don't "get" why it's "needed" to pass a string guard_name when assigning roles/permissions.

Practically speaking, I don't think most implementations bother being aware of which Guard they "intend" to assign a role or permission "for".

In the case of legitimately intending a certain alternate guard, querying the model from the db will be enough to use it when assigning.

And in the majority of cases most apps only really care about a single set of permissions, agnostic of which guard the user is connected via.

**Comment by drbyte** [comment]
In your sample code you're using `->can(PERMISSION_NAME)` instead of `hasPermissionTo(PERMISSION_NAME, GUARD_NAME)` or `hasPermissionTo(PERMISSION_MODEL_INSTANCE)`. 
Using `can()` requires a lookup of the guard_name and uses defaults, whereas with `hasPermissionTo()` you can pass a guard_name or an instance which already relates to the guard_name.

However, if your app structure does NOT differentiate between guards when it comes to roles/permissions, (ie: if ALL your roles/permissions are the SAME for ALL guards), you could try adding this override to your User model, and then you'd only need to create roles/permissions for that single guard_name, not duplicating them:

```php
    protected function getDefaultGuardName(): string { return 'web'; }
```

**Comment by tomekkicaj700** [comment]
Thanks drbyte but still same problem, only the below returns true


```
    $this->actingAs(User::find(1), "api");  //no matter if  guard is "web" or "api" or none
    dd(
      auth()
        ->user()
        ->hasPermissionTo(Permissions::CAN_MANAGE_WELDERS, "web")
    );
returns TRUE
```

`hasPermissionTo(Permissions::CAN_MANAGE_WELDERS, "api")` always returns FALSE

my ROLES table looks like this:

<HTML>
<BODY>
<TABLE>
<TR><td>id</td><td>name</td><td>guard_name</td></TR>
<TR>
<TD>1</TD><TD>Admin</TD><TD>web</TD>
</TR>
<TR>
<TD>2</TD><TD>Admin</TD><TD>api</TD>
</TR>

</TABLE>

</BODY>
</HTML>

my MODEL_HAS_ROLES table looks like this:

<HTML>
<BODY>
<TABLE>
<TR><td>role_id</td><td>model_type</td><td>model_id</td></TR>
<TR>
<TD>1</TD><TD>App\Models\User</TD><TD>1</TD>
</TR>


</TABLE>

</BODY>
</HTML>


Now, I manually editted MODEL_HAS_ROLES  and added new row:

<HTML>
<BODY>
<TABLE>
<TR><td>role_id</td><td>model_type</td><td>model_id</td></TR>
<TR>
<TD>1</TD><TD>App\Models\User</TD><TD>1</TD>
</TR>
<TR>
<TD>2</TD><TD>App\Models\User</TD><TD>1</TD>
</TR>

</TABLE>

</BODY>
</HTML>


..and everything works perfectly, both your suggestion: 
 `hasPermissionTo(Permissions::CAN_MANAGE_WELDERS, "api")` 
and 
`can(Permissions::CAN_MANAGE_WELDERS)`

Looks like I need to call:

`User::find(1)->assignRole(Roles::ADMIN)` with both `"api" `and `"web"` guards but `assignRole `doesn't seem to accept the guard
argument?

**Reply by drbyte** [reply, marked as answer]
$role = Role::findByName(Roles::ADMIN, 'api');

$user->assignRole($role);

Edit: I think you have to pass an attributes array to findByName...

**Reply by tomekkicaj700** [reply]
Many thanks! I didn't think of that! 
Your other sollution with:
`    protected function getDefaultGuardName(): string { return 'web'; }`

works too, I need to figure out yet if it's worthwhile to split permissions into web and api domains.

cheers
Tom
